drivers: display_sdl: support transparent screens

Finomnis opened the pull request:
The `display_sdl` already supports `ARGB8888` colors, but does not display them correctly.

This PR performs the following things:
- Add a gray checkerboard as background to the `display_sdl` driver
- Add the sample `samples/modules/lvgl/screen_transparency` that demonstrates its functionality
- Adjust `samples/drivers/display` in ARGB8888 mode to have a transparent background
- Bug fixes

`screen_transparency` sample before adding transparency support to `display_sdl`:
![image](https://github.com/user-attachments/assets/0d5163b4-c53c-4ef8-b7df-898f73c8cc8e)

with transparency support in `display_sdl`:
![image](https://github.com/user-attachments/assets/25a5edee-1064-4296-9756-b4994a5effba)

**NOTE:** The artifacts around the letters are a bug in LVGL. That bug is fixed in LVGL 9, but in 8.4, this is what it really looks like, this is not a mistake in the `display_sdl` driver.

This is the updated `samples/drivers/display` in `ARGB8888` mode:
![image](https://github.com/user-attachments/assets/ce1c5287-7556-411a-9f44-baf8d831b01d)

Finomnis committed:
samples: modules: lvgl: add `screen_transparency` sample

Adds a sample that renders the simple `hello world` LVGL demo
with a transparent background, to demonstrate ARGB8888 framebuffer
capabilities.

Signed-off-by: Martin Stumpf <finomnis@gmail.com>

Finomnis committed:
drivers: display_sdl: add alpha support

While the driver was already capable of processing `ARGB8888` data,
it did not actually show the alpha value in any way.

This change adds a checkerboard background that shows transparent
regions.

Signed-off-by: Martin Stumpf <finomnis@gmail.com>

Finomnis committed:
samples: modules: lvgl: fix integer overflow

`lv_task_handler()` returns a `uint32_t`, but `k_msleep` takes a
`int32_t`.

If no timer exists, `lv_task_handler()` returns `UINT32_MAX` to indicate
that we should wait forever. However, this gets auto-cast to `-1`,
indicating to `k_msleep` to not wait at all, creating a busy loop.

Hence, a clamping to `[0, INT32_MAX]` is required.

Signed-off-by: Martin Stumpf <finomnis@gmail.com>

kartben commented:
@Finomnis very nice! Could you share a screenshot of how https://github.com/zephyrproject-rtos/zephyr/tree/main/samples%2Fdrivers%2Fdisplay? 
I like the idea of the checkerboard but it feels like it would be nice if the background was customizable, i.e maybe a choice of WHITE/BLACK/GREEN/MAGENTA/CHECKER? And maybe white should still be the default?

Finomnis commented:
> I like the idea of the checkerboard but it feels like it would be nice if the background was customizable, i.e maybe a choice of WHITE/BLACK/GREEN/MAGENTA/CHECKER? And maybe white should still be the default?

Should we make it configurable via Kconfig?

Finomnis commented:
> Could you share a screenshot of how https://github.com/zephyrproject-rtos/zephyr/tree/main/samples%2Fdrivers%2Fdisplay?

![image](https://github.com/user-attachments/assets/45ae147d-00a1-4467-a925-692231fc34bf)
Would you mind explaining? The colors are `0x00ff0000` and so on, so in my opinion it's correct that they are 100% transparent. Is this a bug in the sample?

Finomnis commented:
@kartben I'm very confused. Is the Alpha value 'transparent' in Zephyr defined as `0xff`? Because that's the complete opposite of what it usually is.

LVGL also does "transparent" as `0x00`.

Finomnis commented:
@kartben I can't find the definition of `ARGB8888`. Can you help me with that?

Finomnis commented:
@kartben LVGL's `lv_color_to32` does it the opposite way, though, which we also use in zephyr:

```c
static inline uint32_t lv_color_to32(lv_color_t color)
{
#if LV_COLOR_DEPTH == 1
    if(color.full == 0)
        return 0xFF000000;
    else
        return 0xFFFFFFFF;
#elif LV_COLOR_DEPTH == 8
    lv_color32_t ret;
    LV_COLOR_SET_R32(ret, LV_COLOR_GET_R(color) * 36); /*(2^8 - 1)/(2^3 - 1) = 255/7 = 36*/
    LV_COLOR_SET_G32(ret, LV_COLOR_GET_G(color) * 36); /*(2^8 - 1)/(2^3 - 1) = 255/7 = 36*/
    LV_COLOR_SET_B32(ret, LV_COLOR_GET_B(color) * 85); /*(2^8 - 1)/(2^2 - 1) = 255/3 = 85*/
    LV_COLOR_SET_A32(ret, 0xFF);
    return ret.full;
```

Finomnis commented:
@kartben That one even mixes it in a single function:
```c
void lvgl_set_px_cb_24bit(lv_disp_drv_t *disp_drv, uint8_t *buf, lv_coord_t buf_w, lv_coord_t x,
			  lv_coord_t y, lv_color_t color, lv_opa_t opa)
{
	uint8_t *buf_xy = buf + x * 3U + y * 3U * buf_w;
	lv_color32_t converted_color;

#ifdef CONFIG_LV_COLOR_DEPTH_32
	if (opa != LV_OPA_COVER) {
		lv_color_t mix_color;

		mix_color.ch.red = *buf_xy;
		mix_color.ch.green = *(buf_xy + 1);
		mix_color.ch.blue = *(buf_xy + 2);
		color = lv_color_mix(color, mix_color, opa);
	}
#endif

	converted_color.full = lv_color_to32(color);
	*buf_xy = converted_color.ch.red;
	*(buf_xy + 1) = converted_color.ch.green;
	*(buf_xy + 2) = converted_color.ch.blue;
}
```

(the `mix_color` gets filled with `0x00`, but the `lv_color_mix` fills it with `lv_color_to32` fills it with `0xff`)

Finomnis commented:
@kartben But I think all of that is out of scope for this change to fix. I can fix it in the `display_sdl` driver, but all the other ones need to be fixed in a separate PR I think.

Finomnis committed:
drivers: display_sdl: fix incorrect color conversion

Non-alpha colors were converted to colors with `0x00` alpha, which makes
them fully transparent.

The correct way would be to add a `0xff` alpha, which this change does.

Signed-off-by: Martin Stumpf <finomnis@gmail.com>

Finomnis commented:
@kartben Made grid size and color Kconfig values.

kartben commented:
> @kartben I'm very confused. Is the Alpha value 'transparent' in Zephyr defined as `0xff`? Because that's the complete opposite of what it usually is.
> 
> LVGL also does "transparent" as `0x00`, and so does the rest of the world.

Yes that looks like a bug, they should be 0xff..... indeed

Finomnis committed:
samples: drivers: display: adjust alpha handling

ARGB8888 values all had alpha '0x00' (=transparent).
Changed them to '0xFF' and added a transparency test patch in the middle
of the screen.

Signed-off-by: Martin Stumpf <finomnis@gmail.com>

Finomnis committed:
docs: add release notes for #81184

Added release notes for PR #81184.

Signed-off-by: Martin Stumpf <finomnis@gmail.com>

Finomnis commented:
> maybe a choice of WHITE/BLACK/GREEN/MAGENTA/CHECKER? And maybe white should still be the default?

@kartben I don't like the idea of a white background by default. It leads to bugs that introduce accidentally transparency not being discovered; that's why it's usually something ridiculous.

The defaults I know are either checkerboard (photoshop and the likes) or pink (game engines). Pink specifically to make it as obviously as possible if accidental transparency exists.

I don't like Pink as a default because it's just too obnoxious for the usecases where transparency is intended, like OSD overlays. That's why I think the checkerboard is a good default.

Just for reference, to document my thoughts behind my current PR decisions.

Finomnis commented:
@faxe1008 Just to make you aware of this PR, in case it collides with the lvgl 9.2 integration

Finomnis commented:
Any news/feedback about this? @danieldegrasse for example?

Also, when will the 4.0 release be, roughly, so we can merge stuff again? 
?

Finomnis commented:
@kartben should I create an issue that gets fixed by this change or is it OK if this change exists on its own?

kartben commented:
> @kartben should I create an issue that gets fixed by this change or is it OK if this change exists on its own? 

The latter is fine!

Finomnis commented:
> The latter is fine!

@kartben In that case, is there any way to bump priority?

It's currently really hard to develop apps with screen transparency :/

kartben commented:
> > The latter is fine!
> 
> @kartben In that case, is there any way to bump priority?
> 
> It's currently really hard to develop apps with screen transparency :/

PR has been open for only a week, right as everyone was pretty busy getting 4.0 ready. Please allow some time for reviewers to have a look at it, likely next week, I would say :)

Finomnis commented:
> PR has been open for only a week, right as everyone was pretty busy getting 4.0 ready. Please allow some time for reviewers to have a look at it, likely next week, I would say :)

Oh, I did not realize it's only been open for a week :D guess I'm being too impatient.

Thanks for the response :)

kartben commented on samples/drivers/display/src/main.c:
```diff
@@ -50,16 +51,19 @@ static void fill_buffer_argb8888(enum corner corner, uint8_t grey, uint8_t *buf,
+		color = 0xFF000000u | grey << 16 | grey << 8 | grey;
+		break;
+	case CENTER:
+		color = 0x00FFFFFFu | grey << 24;
```
I think it would be more intuitive for the background to be transparent (`bg_color = 0x00u;`) and the center rectangle being white?

```suggestion
		color = 0xFFFFFFFFu | grey << 24;
```

Finomnis replied:
Sure, we can do that, then I should remove the `grey` fading of the transparency though.

The patch currently fades from transparent to white.

Finomnis replied:
Fixed

Finomnis replied:
![image](https://github.com/user-attachments/assets/ce1c5287-7556-411a-9f44-baf8d831b01d)

kartben commented on drivers/display/Kconfig.sdl:
```diff
@@ -59,4 +59,22 @@ config SDL_DISPLAY_MONO_MSB_FIRST
+	int "Transparency grid cell size"
+	default 8
+	help
+	  The size of a the checkerboard pattern squares, in pixels.
```
```suggestion
	  The size of the checkerboard pattern squares, in pixels.
```

kartben commented on drivers/display/display_sdl_bottom.c:
```diff
@@ -59,16 +62,50 @@ int sdl_display_init_bottom(uint16_t height, uint16_t width, uint16_t zoom_pct,
+
+	err = SDL_LockTexture(*background_texture, NULL, &background_data, &background_pitch);
+	if (err != 0) {
+		nsi_print_warning("Failed to lock background texture: %s", err);
```
```suggestion
		nsi_print_warning("Failed to lock background texture: %d", err);
```

kartben commented on samples/modules/lvgl/screen_transparency/src/main.c:
```diff
@@ -0,0 +1,73 @@
+	/* Create a label, set its text and align it to the center */
+	label = lv_label_create(lv_scr_act());
+	lv_label_set_text(label, "Hello, world!");
+	lv_obj_set_style_text_color(lv_scr_act(), lv_color_hex(0xff00ff), LV_PART_MAIN);
```
shouldn't this be 
```suggestion
	lv_obj_set_style_text_color(label, lv_color_hex(0xff00ff), LV_PART_MAIN);
```

Finomnis replied:
That's copy paste from https://github.com/lvgl/lvgl/blob/master/examples/get_started/lv_example_get_started_1.c, but, maybe? I'm not fully familiar with LVGL yet, I'm only setting up the tooling for it for our project right now.

Finomnis replied:
But I would probably keep it as close as possible to the lvgl hello world example

Finomnis commented:
@kartben Thanks for the review and your time :)
